openmdao/core/implicitcomponent.py

```diff
@@ -105,21 +133,46 @@ def _apply_linear(self, vec_names, mode, scope_out=None, scope_in=None):
             If None, all are in the scope.
         """
         for vec_name in vec_names:
+            if vec_name not in self._rel_vec_names:
+                continue
+
             with self._matvec_context(vec_name, scope_out, scope_in, mode) as vecs:
                 d_inputs, d_outputs, d_residuals = vecs
 
                 # Jacobian and vectors are all scaled, unitless
                 with self.jacobian_context() as J:
                     J._apply(d_inputs, d_outputs, d_residuals, mode)
 
+                # if we're not matrix free, we can skip the bottom of
+                # this loop because apply_linear does nothing.
+                if not self.matrix_free:
```

Comment:
same comment as with explicit component. Shouldn't do the apply if were' matrix free...